src/main/java/frc/robot/subsystems/shooter/commands/SpeedUp.java

```diff
@@ -0,0 +1,28 @@
+package frc.robot.subsystems.shooter.commands;
+
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import frc.robot.subsystems.shooter.Shooter;
+
+/**
+ * This command changes the velocity of the shooter to a specified velocity, in order to shoot at the best rate.
+ */
+public class SpeedUp extends CommandBase {
+    private final Shooter shooter;
+    private final double velocity;
+
+    public SpeedUp(Shooter shooter, double velocity) {
```

Comment:
I think this command should receive the distance and calculate the velocity in the execute(), in case the robot moves. The Shoot command may also need this.

Reply:
The only way I could do it, is using something similar to `Supplier<T>`, because if I won't do it this way, I won't be able to know what is the current distance. Secondly, If I want to do it, I need to calculate the velocity in a different way, because it's a relative speed, and we haven't tested it how it work. I think we should stay using velocity, because it would be hard to change & test during the season.

Reply:
Okay, so what you are saying is that this command will change after testing during the season?

Reply:
Oh wait, I think I understand, what if the vision outputs the distance in the execute()?

Reply:
Okay so if I'm not wrong this is what should happen:
- There should be a command for testing that receives the velocity, without all the distance calculations.
- Then there should be SpeedUp that in the execute takes the distance from the Vision, calculates the velocity and stops when it gets to the calculated velocity.
- Shoot command activates and actually shoots the powercell

Reply:
For the first Question, The command may change, we would have enough time. You need to remember that the shooter was very bad in lowering the speed, most likely it would happen again.
For the second Question, we can do it, but it's not very modular this way, I think using supplier would be better.

Reply:
For the second bullet a supplier would work as well.

Reply:
I'll put my two agoras by combining a few of the things you two have said:
> I think we should stay using velocity

I agree. IMO, this command should receive a velocity setpoint. It doesn't matter where the setpoint came from: sent over Network Tables, gotten from vision calculations, Kalman or Our Lord And Saviour Deam Kamen entered the roboRIO and inputted the value. It's much simpler that way because you have one command that supports every setpoint outputter, future ones included.
> The only way I could do it, is using something similar to `Supplier<T>`

You should still use suppliers, otherwise (as Barel said) you wouldn't be able to change the setpoint while shooting. If you want to test a single value, you can still do that (`() -> 300`), but the opposite can't be done without suppliers. This is called the _[single responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle)_.

> Then there should be SpeedUp that in the execute takes the distance from the Vision, calculates the velocity and stops when it gets to the calculated velocity.

I imagined something similar.
However, I believe your final goal is to calculate the distance using localization, isn't it? That way you'll be able to shoot accurately (maybe, will need testing) even when the vision target isn't visible.

Reply:
We will get the distance using localization, in combination with the vision. During the season we will test which one is more accurate, and then we use the more accurate one. About the part that the vision target isn't visible, if I understand you, the shooter won't be in the correct rotation, because this year we don't have Turret, so the shooter is aligned with the drivetrain.

Reply:
> About the part that the vision target isn't visible, if I understand you, the shooter won't be in the correct rotation, because this year we don't have Turret, so the shooter is aligned with the drivetrain.

I know there won't be a turret, but I don't see how it's relevant to the vision target's visibility.

Reply:
I think that there is a function for checking if the vision sees the target, so we can determine by the case we won't see the target and to act appropriately.